Extend SCC construction to enable extra functionality

amandasystems opened the pull request:
Do YOU feel like your SCC construction doesn't do enough? Then I have a patch for you! SCCs can now do *everything*! Well, almost.

This patch has been extracted from #123720. It specifically enhances
`Sccs` to allow tracking arbitrary commutative properties (think min/max mappings on nodes vs arbitrary closures) of strongly connected components, including
- reachable values (max/min)
- SCC-internal values (max/min)

This helps with among other things universe computation. We can now identify
SCC universes as a reasonably straightforward "find max/min" operation during SCC construction. This is also included in this patch.

It's also more or less zero-cost; don't use the new features, don't pay for them.

This commit also vastly extends the documentation of the SCCs module, which I had a very hard time following. It may or may not have gotten easier to read for someone else.

I believe this logic can also be used in leak check, but haven't checked. Ha. ha. Ha.

amandasystems committed:
Extend SCC construction to enable extra functionality

This patch has been extracted from #123720. It specifically enhances
`Sccs` to allow tracking arbitrary commutative properties of SCCs, including
- reachable values (max/min)
- SCC-internal values (max/min)

This helps with among other things universe computation: we can now identify
SCC universes as a straightforward "find max/min" operation during SCC construction.

It's also more or less zero-cost; don't use the new features, don't pay for them.

This commit also vastly extends the documentation of the SCCs module, which I had a very hard time following.

amandasystems committed:
Documentation fixes

amandasystems commented:
r? Lqd

lqd commented:
@bors try @rust-timer queue

As discussed on zulip, until I have more time let's also temporarily redirect to niko for review or reassignment, and discuss at this week's meeting: r? nikomatsakis

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 609a202928a8ad7a29f7ea26c4101c2c176b6b1d with merge 6c7f0a6914780a8bd5240c26e668f7b0625b66dc...
<!-- homu: {"type":"TryBuildStarted","head_sha":"609a202928a8ad7a29f7ea26c4101c2c176b6b1d","merge_sha":"6c7f0a6914780a8bd5240c26e668f7b0625b66dc"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9063491717/job/24904485731)
Build commit: 6c7f0a6914780a8bd5240c26e668f7b0625b66dc (`6c7f0a6914780a8bd5240c26e668f7b0625b66dc`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9063491717/job/24904485731"},"merge_sha":"6c7f0a6914780a8bd5240c26e668f7b0625b66dc"} -->

rust-timer commented:
Queued 6c7f0a6914780a8bd5240c26e668f7b0625b66dc with parent 6be7b0c7d2b085474f9f2f9323c2266f4df105d8, future [comparison URL](https://perf.rust-lang.org/compare.html?start=6be7b0c7d2b085474f9f2f9323c2266f4df105d8&end=6c7f0a6914780a8bd5240c26e668f7b0625b66dc).
There are currently 0 preceding artifacts in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.2 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (6c7f0a6914780a8bd5240c26e668f7b0625b66dc): [comparison URL](https://perf.rust-lang.org/compare.html?start=6be7b0c7d2b085474f9f2f9323c2266f4df105d8&end=6c7f0a6914780a8bd5240c26e668f7b0625b66dc&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=6be7b0c7d2b085474f9f2f9323c2266f4df105d8&end=6c7f0a6914780a8bd5240c26e668f7b0625b66dc&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.5% | [0.2%, 1.6%] | 45    |
| Regressions ❌ <br /> (secondary)  | 1.4% | [0.1%, 4.0%] | 32    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.5% | [0.2%, 1.6%] | 45    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=6be7b0c7d2b085474f9f2f9323c2266f4df105d8&end=6c7f0a6914780a8bd5240c26e668f7b0625b66dc&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.8%  | [2.4%, 3.2%]   | 2     |
| Regressions ❌ <br /> (secondary)  | 2.2%  | [2.2%, 2.2%]   | 2     |
| Improvements ✅ <br /> (primary)   | -2.6% | [-2.6%, -2.6%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 1.0%  | [-2.6%, 3.2%]  | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=6be7b0c7d2b085474f9f2f9323c2266f4df105d8&end=6c7f0a6914780a8bd5240c26e668f7b0625b66dc&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 1.9%  | [1.6%, 2.1%]   | 3     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.9% | [-2.9%, -2.9%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=6be7b0c7d2b085474f9f2f9323c2266f4df105d8&end=6c7f0a6914780a8bd5240c26e668f7b0625b66dc&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 675.291s -> 676.941s (0.24%)
**Artifact size**: 315.98 MiB -> 316.29 MiB (0.10%)

amandasystems committed:
Simplify path compression logic

amandasystems committed:
Formatting, weird because I just did that

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit c510e73513a47bc17bc52971e425cb3c929af4fb with merge 1805d8ae3b6d155ffdc3ee3371389de5042fb85e...
<!-- homu: {"type":"TryBuildStarted","head_sha":"c510e73513a47bc17bc52971e425cb3c929af4fb","merge_sha":"1805d8ae3b6d155ffdc3ee3371389de5042fb85e"} -->

amandasystems committed:
Docstring for for `Annotation`

Note that this changes no executing code. The change is 100% in documentation.

amandasystems commented:
My guess is 0.3% primary regressions and 0.84% in secondary, let's see how wrong I am!

lqd commented:
@rust-timer build 1805d8ae3b6d155ffdc3ee3371389de5042fb85e

rust-timer commented:
Queued 1805d8ae3b6d155ffdc3ee3371389de5042fb85e with parent a5c37eea5aa922f4d6b543f2d35bdbd892fea2a8, future [comparison URL](https://perf.rust-lang.org/compare.html?start=a5c37eea5aa922f4d6b543f2d35bdbd892fea2a8&end=1805d8ae3b6d155ffdc3ee3371389de5042fb85e).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~2.0 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (1805d8ae3b6d155ffdc3ee3371389de5042fb85e): [comparison URL](https://perf.rust-lang.org/compare.html?start=a5c37eea5aa922f4d6b543f2d35bdbd892fea2a8&end=1805d8ae3b6d155ffdc3ee3371389de5042fb85e&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=a5c37eea5aa922f4d6b543f2d35bdbd892fea2a8&end=1805d8ae3b6d155ffdc3ee3371389de5042fb85e&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3% | [0.2%, 0.8%] | 23    |
| Regressions ❌ <br /> (secondary)  | 1.5% | [0.1%, 3.9%] | 20    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.3% | [0.2%, 0.8%] | 23    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=a5c37eea5aa922f4d6b543f2d35bdbd892fea2a8&end=1805d8ae3b6d155ffdc3ee3371389de5042fb85e&stat=max-rss)
<details>
<summary>Results (primary -0.4%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.3%  | [1.3%, 1.3%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -1.3% | [-2.4%, -0.2%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.4% | [-2.4%, 1.3%]  | 3     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=a5c37eea5aa922f4d6b543f2d35bdbd892fea2a8&end=1805d8ae3b6d155ffdc3ee3371389de5042fb85e&stat=cycles:u)
<details>
<summary>Results (primary -0.2%, secondary -1.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 7.2%  | [7.2%, 7.2%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 2.0%  | [1.8%, 2.2%]   | 5     |
| Improvements ✅ <br /> (primary)   | -2.1% | [-4.3%, -0.8%] | 4     |
| Improvements ✅ <br /> (secondary) | -2.6% | [-3.8%, -2.0%] | 10    |
| All ❌✅ (primary)                 | -0.2% | [-4.3%, 7.2%]  | 5     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=a5c37eea5aa922f4d6b543f2d35bdbd892fea2a8&end=1805d8ae3b6d155ffdc3ee3371389de5042fb85e&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 669.422s -> 671.224s (0.27%)
**Artifact size**: 316.04 MiB -> 316.26 MiB (0.07%)

amandasystems committed:
Slightly faster version of `find_state`

This version shaves off ca 2% of the cycles in my experiments
and makes the control flow easier to follow for me and hopefully
others, including the compiler.

Someone gave me a working profiler and by God I'm using it.

amandasystems committed:
Remove a few unnecessary constructions

This shaves off ca 6% of the cycles in `start_walk_from()` in my
experiments.

Kobzol commented:
@bors try @rust-timer queue

bors commented:
:hourglass: Trying commit b453158e337396ddbf0439079ba26c831086feb1 with merge f02f9b5f83a84a26d250b683819145c63ef1004c...
<!-- homu: {"type":"TryBuildStarted","head_sha":"b453158e337396ddbf0439079ba26c831086feb1","merge_sha":"f02f9b5f83a84a26d250b683819145c63ef1004c"} -->

jackh726 commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit b453158e337396ddbf0439079ba26c831086feb1 with merge f3129a502c9cb5f245ac39b47c8a169b68fba662...
<!-- homu: {"type":"TryBuildStarted","head_sha":"b453158e337396ddbf0439079ba26c831086feb1","merge_sha":"f3129a502c9cb5f245ac39b47c8a169b68fba662"} -->

Kobzol commented:
Hmm, looks like we started two try runs in a row, but the previous one was not cancelled. That looks like a CI config bug (?).

jackh726 commented:
Oops, I missed that you started a try/perf run before I did...

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9159502031/job/25184289038)
Build commit: f3129a502c9cb5f245ac39b47c8a169b68fba662 (`f3129a502c9cb5f245ac39b47c8a169b68fba662`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9159502031/job/25184289038"},"merge_sha":"f3129a502c9cb5f245ac39b47c8a169b68fba662"} -->

rust-timer commented:
Queued f3129a502c9cb5f245ac39b47c8a169b68fba662 with parent e8ada6ab253b510ac88edda131021d9878f2984f, future [comparison URL](https://perf.rust-lang.org/compare.html?start=e8ada6ab253b510ac88edda131021d9878f2984f&end=f3129a502c9cb5f245ac39b47c8a169b68fba662).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.8 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (f3129a502c9cb5f245ac39b47c8a169b68fba662): [comparison URL](https://perf.rust-lang.org/compare.html?start=e8ada6ab253b510ac88edda131021d9878f2984f&end=f3129a502c9cb5f245ac39b47c8a169b68fba662&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=e8ada6ab253b510ac88edda131021d9878f2984f&end=f3129a502c9cb5f245ac39b47c8a169b68fba662&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.3% | [0.2%, 0.7%] | 26    |
| Regressions ❌ <br /> (secondary)  | 1.3% | [0.1%, 3.9%] | 21    |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.3% | [0.2%, 0.7%] | 26    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=e8ada6ab253b510ac88edda131021d9878f2984f&end=f3129a502c9cb5f245ac39b47c8a169b68fba662&stat=max-rss)
<details>
<summary>Results (primary -0.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.4%  | [2.4%, 2.4%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.6% | [-2.6%, -2.6%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.1% | [-2.6%, 2.4%]  | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=e8ada6ab253b510ac88edda131021d9878f2984f&end=f3129a502c9cb5f245ac39b47c8a169b68fba662&stat=cycles:u)
<details>
<summary>Results (primary 2.4%, secondary 1.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.4%  | [2.4%, 2.4%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 2.2%  | [2.1%, 2.5%]   | 8     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.0% | [-3.9%, -2.0%] | 2     |
| All ❌✅ (primary)                 | 2.4%  | [2.4%, 2.4%]   | 1     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=e8ada6ab253b510ac88edda131021d9878f2984f&end=f3129a502c9cb5f245ac39b47c8a169b68fba662&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 670.321s -> 670.92s (0.09%)
**Artifact size**: 316.18 MiB -> 316.07 MiB (-0.03%)

lqd commented on compiler/rustc_borrowck/src/region_infer/mod.rs:
```diff
@@ -72,7 +72,7 @@ pub struct RegionInferenceContext<'tcx> {
     /// graph. We have an edge from SCC A to SCC B if `A: B`. Used to
     /// compute the values of each region.
-    constraint_sccs: Rc<Sccs<RegionVid, ConstraintSccIndex>>,
+    pub constraint_sccs: Rc<ConstraintSccs>,
```
do we need this pub since we already have the `constraint_sccs` method?

lqd commented on compiler/rustc_data_structures/src/graph/scc/mod.rs:
```diff
@@ -4,52 +4,119 @@
 /// the SCCs. We can map from each node to the SCC that it
 /// participates in, and we also have the successors of each SCC.
-pub struct Sccs<N: Idx, S: Idx> {
+pub struct Sccs<N: Idx, S: Idx, A: Annotation> {
```
would A having a default of `()` work here?

lqd commented on compiler/rustc_data_structures/src/graph/scc/mod.rs:
```diff
@@ -225,24 +314,28 @@ where
     /// D' (i.e., D' < D), we know that N, N', and all nodes in
     /// between them on the stack are part of an SCC.
     ///
+    /// Additionally, we keep track of a representative annotation of the
```
Maybe reusing the word "representative" here could be confusing?

lqd commented on compiler/rustc_borrowck/src/region_infer/mod.rs:
```diff
@@ -91,22 +91,6 @@ pub struct RegionInferenceContext<'tcx> {
     /// Map universe indexes to information on why we created it.
     universe_causes: FxIndexMap<ty::UniverseIndex, UniverseInfo<'tcx>>,
 
-    /// Contains the minimum universe of any variable within the same
-    /// SCC. We will ensure that no SCC contains values that are not
-    /// visible from this index.
-    scc_universes: IndexVec<ConstraintSccIndex, ty::UniverseIndex>,
-
-    /// Contains the "representative" region of each SCC.
-    /// It is defined as the one with the minimal RegionVid, favoring
-    /// free regions, then placeholders, then existential regions.
-    ///
-    /// It is a hacky way to manage checking regions for equality,
-    /// since we can 'canonicalize' each region to the representative
-    /// of its SCC and be sure that -- if they have the same repr --
-    /// they *must* be equal (though not having the same repr does not
-    /// mean they are unequal).
-    scc_representatives: IndexVec<ConstraintSccIndex, ty::RegionVid>,
-
```
From a quick read I've not found where these comments moved to, did we lose some maybe?

lqd commented on compiler/rustc_borrowck/src/region_infer/mod.rs:
````diff
@@ -391,123 +377,6 @@ impl<'tcx> RegionInferenceContext<'tcx> {
-            }
-        }
-
-        // Walk each SCC `A` and `B` such that `A: B`
-        // and ensure that universe(A) can see universe(B).
-        //
-        // This serves to enforce the 'empty/placeholder' hierarchy
-        // (described in more detail on `RegionKind`):
-        //
-        // ```
-        // static -----+
-        //   |         |
-        // empty(U0) placeholder(U1)
-        //   |      /
-        // empty(U1)
-        // ```
-        //
-        // In particular, imagine we have variables R0 in U0 and R1
-        // created in U1, and constraints like this;
-        //
-        // ```
-        // R1: !1 // R1 outlives the placeholder in U1
-        // R1: R0 // R1 outlives R0
-        // ```
-        //
-        // Here, we wish for R1 to be `'static`, because it
-        // cannot outlive `placeholder(U1)` and `empty(U0)` any other way.
-        //
-        // Thanks to this loop, what happens is that the `R1: R0`
-        // constraint lowers the universe of `R1` to `U0`, which in turn
-        // means that the `R1: !1` constraint will (later) cause
-        // `R1` to become `'static`.
````
Same question for all the comments in this function, we kept all this somewhere else, right? Or are they not applicable anymore?

lqd commented on compiler/rustc_data_structures/src/graph/scc/mod.rs:
```diff
@@ -4,52 +4,119 @@
+///
+/// Concretely, the following properties must hold (where `merge`
+/// is `merge_scc` and `merge_reached`):
+/// - idempotency: `a.merge(a) = a`
```
I guess we don't need to mention transitivity because it's a given in this context?

In which cases is an annotation merged with itself?

lqd commented on compiler/rustc_borrowck/src/region_infer/mod.rs:
```diff
@@ -2297,6 +2162,16 @@ impl<'tcx> RegionDefinition<'tcx> {
         Self { origin, universe, external_name: None }
     }
+
+    #[inline(always)]
+    pub fn is_placeholder(&self) -> bool {
+        matches!(self.origin, NllRegionVariableOrigin::Placeholder(_))
+    }
+
+    #[inline(always)]
+    pub fn is_existential(&self) -> bool {
+        matches!(self.origin, NllRegionVariableOrigin::Existential { .. })
+    }
```
are these used anywhere?

amandasystems committed:
Roll back a few `#[instrument]`

Apparently this interferes with inlining and murders performance
on `wg-grammar`.

amandasystems committed:
Resurrect some comments, remove a `pub`

This commit addresses @lqd's code review and resurrects
a lost comment and removes some dead code.

amandasystems committed:
Revise documentation after @lqd's comments

amandasystems committed:
Move `RegionTracker` to `region_infer`

In terms of code organisation, this is a lot cleaner and allows
tighter access modifiers.

lqd commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 58ccdd09cbe0cdfc953ea81113b941105601c899 with merge 4580462219447558f8ed6b6c8e1baa94d6fc24f8...
<!-- homu: {"type":"TryBuildStarted","head_sha":"58ccdd09cbe0cdfc953ea81113b941105601c899","merge_sha":"4580462219447558f8ed6b6c8e1baa94d6fc24f8"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9189551797/job/25276013391)
Build commit: 4580462219447558f8ed6b6c8e1baa94d6fc24f8 (`4580462219447558f8ed6b6c8e1baa94d6fc24f8`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9189551797/job/25276013391"},"merge_sha":"4580462219447558f8ed6b6c8e1baa94d6fc24f8"} -->

rust-timer commented:
Queued 4580462219447558f8ed6b6c8e1baa94d6fc24f8 with parent b54dd08a84f3c07efbc2aaf63c3df219ae680a03, future [comparison URL](https://perf.rust-lang.org/compare.html?start=b54dd08a84f3c07efbc2aaf63c3df219ae680a03&end=4580462219447558f8ed6b6c8e1baa94d6fc24f8).
There is currently 1 preceding artifact in the [queue](https://perf.rust-lang.org/status.html).
It will probably take at least ~1.3 hours until the benchmark run finishes.
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (4580462219447558f8ed6b6c8e1baa94d6fc24f8): [comparison URL](https://perf.rust-lang.org/compare.html?start=b54dd08a84f3c07efbc2aaf63c3df219ae680a03&end=4580462219447558f8ed6b6c8e1baa94d6fc24f8&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=b54dd08a84f3c07efbc2aaf63c3df219ae680a03&end=4580462219447558f8ed6b6c8e1baa94d6fc24f8&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.2% | [0.2%, 0.3%] | 2     |
| Regressions ❌ <br /> (secondary)  | 0.6% | [0.2%, 1.1%] | 8     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.2% | [0.2%, 0.3%] | 2     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=b54dd08a84f3c07efbc2aaf63c3df219ae680a03&end=4580462219447558f8ed6b6c8e1baa94d6fc24f8&stat=max-rss)
<details>
<summary>Results (primary -2.3%, secondary -3.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.3% | [-2.3%, -2.3%] | 1     |
| Improvements ✅ <br /> (secondary) | -3.0% | [-3.2%, -2.9%] | 2     |
| All ❌✅ (primary)                 | -2.3% | [-2.3%, -2.3%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=b54dd08a84f3c07efbc2aaf63c3df219ae680a03&end=4580462219447558f8ed6b6c8e1baa94d6fc24f8&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=b54dd08a84f3c07efbc2aaf63c3df219ae680a03&end=4580462219447558f8ed6b6c8e1baa94d6fc24f8&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 672.666s -> 672.967s (0.04%)
**Artifact size**: 315.47 MiB -> 316.17 MiB (0.22%)

nikomatsakis commented:
@bors r+

bors commented:
:pushpin: Commit 58ccdd09cbe0cdfc953ea81113b941105601c899 has been approved by `nikomatsakis`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=nikomatsakis 58ccdd09cbe0cdfc953ea81113b941105601c899 -->
<!-- homu: {"type":"Approved","sha":"58ccdd09cbe0cdfc953ea81113b941105601c899","approver":"nikomatsakis","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

nikomatsakis commented:
Actually that was slightly premature. Let me finish the read :)

@bors r-

nikomatsakis commented on compiler/rustc_data_structures/src/graph/scc/mod.rs:
```diff
@@ -4,52 +4,116 @@
+
+// The name of this struct should discourage you from making it public and leaking
+// its representation. This message was left here by one who came before you,
+//  who learnt the hard way that making even small changes in representation is difficult when it's publicly inspectable. Obey the law of Demeter!
```
🤣

lqd replied:
Interestingly, tidy isn't pointing out this line as being over our 100 character-limit as it usually does.

nikomatsakis reviewed (commented):
r=me modulo nits

I love all the comments you added! 💜

amandasystems committed:
Address code review comments on the comments

amandasystems commented:
Rebased!

jackh726 commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit d63708b9074ebdc69e21bee8588d28db497e9ccd has been approved by `nikomatsakis`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=nikomatsakis d63708b9074ebdc69e21bee8588d28db497e9ccd -->
<!-- homu: {"type":"Approved","sha":"d63708b9074ebdc69e21bee8588d28db497e9ccd","approver":"nikomatsakis","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit d63708b9074ebdc69e21bee8588d28db497e9ccd with merge 8cf5101d77cd9eeb12751c563d8098aba2c604d0...
<!-- homu: {"type":"BuildStarted","head_sha":"d63708b9074ebdc69e21bee8588d28db497e9ccd","merge_sha":"8cf5101d77cd9eeb12751c563d8098aba2c604d0"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9491010544/job/26159540800)
Approved by: nikomatsakis
Pushing 8cf5101d77cd9eeb12751c563d8098aba2c604d0 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nikomatsakis","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9491010544/job/26159540800"},"merge_sha":"8cf5101d77cd9eeb12751c563d8098aba2c604d0"} -->

rust-timer commented:
Finished benchmarking commit (8cf5101d77cd9eeb12751c563d8098aba2c604d0): [comparison URL](https://perf.rust-lang.org/compare.html?start=8337ba9189de188e2ed417018af2bf17a57d51ac&end=8cf5101d77cd9eeb12751c563d8098aba2c604d0&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


**Next Steps**: If you can justify the regressions found in this perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please open an issue or create a new PR that fixes the regressions, add a comment linking to the newly created issue or PR, and then add the `perf-regression-triaged` label to this PR.

@rustbot label: +perf-regression
cc @rust-lang/wg-compiler-performance



### [Instruction count](https://perf.rust-lang.org/compare.html?start=8337ba9189de188e2ed417018af2bf17a57d51ac&end=8cf5101d77cd9eeb12751c563d8098aba2c604d0&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.2%  | [0.2%, 0.3%]   | 6     |
| Regressions ❌ <br /> (secondary)  | 0.5%  | [0.1%, 1.0%]   | 10    |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.3% | [-0.3%, -0.3%] | 2     |
| All ❌✅ (primary)                 | 0.2%  | [0.2%, 0.3%]   | 6     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=8337ba9189de188e2ed417018af2bf17a57d51ac&end=8cf5101d77cd9eeb12751c563d8098aba2c604d0&stat=max-rss)
<details>
<summary>Results (primary -3.2%, secondary -5.0%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -3.2% | [-4.4%, -2.0%] | 2     |
| Improvements ✅ <br /> (secondary) | -5.0% | [-5.0%, -5.0%] | 1     |
| All ❌✅ (primary)                 | -3.2% | [-4.4%, -2.0%] | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=8337ba9189de188e2ed417018af2bf17a57d51ac&end=8cf5101d77cd9eeb12751c563d8098aba2c604d0&stat=cycles:u)
<details>
<summary>Results (secondary -1.8%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.9%  | [2.0%, 3.6%]   | 4     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -8.1% | [-9.1%, -7.5%] | 3     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=8337ba9189de188e2ed417018af2bf17a57d51ac&end=8cf5101d77cd9eeb12751c563d8098aba2c604d0&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 671.52s -> 672.648s (0.17%)
**Artifact size**: 320.35 MiB -> 320.29 MiB (-0.02%)

rylev commented:
@amandasystems @nikomatsakis I see that the perf regressions were brought down since the first perf run, but there's still instruction count regressions in primary benchmarks here (albeit not large ones). Shall we just mark this as triaged and move on?

amandasystems commented:
> Shall we just mark this as triaged and move on?

That’s how I understood the decision anyway!
